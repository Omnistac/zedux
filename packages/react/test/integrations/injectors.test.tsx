import {
  atom,
  createEcosystem,
  injectCallback,
  injectEffect,
  injectMemo,
  injectRef,
  injectStore,
} from '@zedux/react'

const ecosystem = createEcosystem({ id: 'test-injectors' })

afterEach(() => {
  ecosystem.reset()
})

describe('injectors', () => {
  test('React-esque injectors mimic React hook functionality', () => {
    const ref = {}
    const cbs: (() => void)[] = []
    const cleanups: string[] = []
    const effects: string[] = []
    const refs: typeof ref[] = []
    const vals: string[] = []
    const cbA = () => {}
    const cbB = () => {}

    const atom1 = atom('1', () => {
      const store = injectStore('a')
      const val1 = injectMemo(() => store.getState())
      const val2 = injectMemo(() => store.getState(), [])
      const val3 = injectMemo(() => store.getState(), [store.getState()])
      vals.push(val1, val2, val3)

      const injectedRef = injectRef(ref)
      refs.push(injectedRef.current)

      const cb1 = injectCallback(store.getState() === 'a' ? cbA : cbB)
      const cb2 = injectCallback(store.getState() === 'a' ? cbA : cbB, [])
      const cb3 = injectCallback(store.getState() === 'a' ? cbA : cbB, [
        store.getState(),
      ])
      cbs.push(cb1, cb2, cb3)

      injectEffect(() => {
        effects.push(store.getState())

        return () => cleanups.push(store.getState())
      }, [store.getState()])

      return store
    })

    const instance = ecosystem.getInstance(atom1)

    instance.setState('b')

    expect(vals).toEqual(['a', 'a', 'a', 'b', 'a', 'b'])
    expect(cbs).toEqual([cbA, cbA, cbA, cbB, cbA, cbB])
    expect(effects).toEqual(['b'])
    expect(cleanups).toEqual([])
    expect(refs).toEqual([ref, ref])

    instance.setState('c')

    expect(vals).toEqual(['a', 'a', 'a', 'b', 'a', 'b', 'c', 'a', 'c'])
    expect(cbs).toEqual([cbA, cbA, cbA, cbB, cbA, cbB, cbB, cbA, cbB])
    expect(effects).toEqual(['b', 'c'])
    expect(cleanups).toEqual(['c'])
    expect(refs).toEqual([ref, ref, ref])
  })
})
