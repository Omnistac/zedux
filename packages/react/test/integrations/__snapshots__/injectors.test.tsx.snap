// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`injectors when called with no deps, injectMemo() tracks used signals 1`] = `
{
  "1": {
    "className": "AtomInstance",
    "observers": {},
    "sources": {
      "@memo(1)-3": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
      "@signal()-2": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "state": 3,
    "status": "Active",
    "weight": 4,
  },
  "@memo(1)-3": {
    "className": "SelectorInstance",
    "observers": {
      "1": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "sources": {
      "@signal()-1": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "state": 1,
    "status": "Active",
    "weight": 2,
  },
  "@signal()-1": {
    "className": "Signal",
    "observers": {
      "@memo(1)-3": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "sources": {},
    "state": 1,
    "status": "Active",
    "weight": 1,
  },
  "@signal()-2": {
    "className": "Signal",
    "observers": {
      "1": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "sources": {},
    "state": 2,
    "status": "Active",
    "weight": 1,
  },
}
`;

exports[`injectors when called with no deps, injectMemo() tracks used signals 2`] = `
{
  "1": {
    "className": "AtomInstance",
    "observers": {},
    "sources": {
      "@memo(1)-3": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
      "@signal()-2": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "state": 13,
    "status": "Active",
    "weight": 4,
  },
  "@memo(1)-3": {
    "className": "SelectorInstance",
    "observers": {
      "1": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "sources": {
      "@signal()-1": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "state": 11,
    "status": "Active",
    "weight": 2,
  },
  "@signal()-1": {
    "className": "Signal",
    "observers": {
      "@memo(1)-3": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "sources": {},
    "state": 11,
    "status": "Active",
    "weight": 1,
  },
  "@signal()-2": {
    "className": "Signal",
    "observers": {
      "1": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "sources": {},
    "state": 2,
    "status": "Active",
    "weight": 1,
  },
}
`;

exports[`injectors when called with no deps, injectMemo() tracks used signals 3`] = `
{
  "1": {
    "className": "AtomInstance",
    "observers": {},
    "sources": {
      "@memo(1)-3": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
      "@signal()-2": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "state": 33,
    "status": "Active",
    "weight": 4,
  },
  "@memo(1)-3": {
    "className": "SelectorInstance",
    "observers": {
      "1": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "sources": {
      "@signal()-1": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "state": 11,
    "status": "Active",
    "weight": 2,
  },
  "@signal()-1": {
    "className": "Signal",
    "observers": {
      "@memo(1)-3": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "sources": {},
    "state": 11,
    "status": "Active",
    "weight": 1,
  },
  "@signal()-2": {
    "className": "Signal",
    "observers": {
      "1": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "sources": {},
    "state": 22,
    "status": "Active",
    "weight": 1,
  },
}
`;

exports[`injectors when called with no deps, injectMemo() tracks used signals 4`] = `{}`;
